### PR TITLE
API proposal

### PR DESCRIPTION
**Pull request by pepe**
The biggest change is to use struct in the `ecs/def-system` and `fsm/define` binding. In both cases, we are transforming tuples into a struct in the macro anyway.

I changed all the examples to this new style and made some of them a little more involved for my learning purposes.

I also formatted the code.

### CODE REVIEW

**Comment by AlecTroemel** [comment]
thanks again for working on this! Some stray thoughts

- for the `everything` example, I like the how the messages content is the key for state transition function. Fun idea!
- changes to FSM are good, definitely cleans up the main macro.
- one thought with `ecs/def-system` is that if we used `[]` instead of `{}` we would mimic some the syntax of `let`, (and `defn` kinda).
    ```clojure
    (let 
      [moveables [:position :velocity]]
      (print "wow"))
      
    (defn contrived
      [moveables] ;; missing "value" part.. so only half counts
      (print "very alike"))

    (def-system move
      [moveables [:position :velocity]]
      (print "so similar))
    ```
  but then we're back to tuples.. So I don't think its a hill to die on. I'm stilling learning Janet's standard lib, do you know of anything else that uses the proposed syntax? specifically this `{moveables [:position :velocity]}`. Again its not super important to me, but I want to stay within Janet norms where it makes sense.

File: examples/05-fsm.janet [hunk 1]
```diff
@@ -1,26 +1,27 @@
...
-  :leave (fn [self] (print "leaving red"))
-  :calm  |(:goto $ :yellow)))
+  colored-warnings
+  {:green
+   {:enter (fn [self] (print "entering green"))
+    :leave (fn [self] (print "leaving green"))
+    :warn (fn [self]
+            (print "before warn")
+            (:goto self :yellow)
+            (print "after warn") :)}
```
Comment (AlecTroemel) [inline examples/05-fsm.janet:17]: is the `:` at the end here a typo?

Comment (pepe) [reply examples/05-fsm.janet:17]: Oh, yes, That is the typo. I will fix it. Kakoune multiselections I blame for this.

**Comment by pepe** [comment]
For the struct vs. tuple in the def- I guess I like to get s.. done in the smallest effort. I was thinking a lot about this and almost removed it, but then I said I would try it with you :-). Again, if you choose not to incorporate these changes, I will still be happy.

One more thing that came to my mind is to remove `-` from `def-` akin to `defmacro`.

**Comment by AlecTroemel** [comment]
Ya I've been thinking about it more and decided `{}` vs `[]` make very little difference, so I'll take the one with the simpler implementation!

**Comment by pepe** [comment]
Thank you very much!

**Pull request merged**